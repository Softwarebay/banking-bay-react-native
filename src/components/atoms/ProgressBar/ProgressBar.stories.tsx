import { color, number } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react-native';
import { ViewStory } from 'utils/storybook/ViewStory';

import { ProgressBar } from './ProgressBar';

storiesOf('Atoms/ProgressBar', module)
	.addDecorator((getStory) => <ViewStory>{getStory()}</ViewStory>)
	.add('Not full', () => (
		<ProgressBar
			maxValue={number('maxValue', 1000)}
			progressColor={color('progressColor', '#3EB290')}
			testID="ProgressBar-test-id"
			value={number('value', 700, { range: true, max: 1000, min: 0, step: 1 })}
		/>
	))
	.add('Full', () => (
		<ProgressBar
			maxValue={number('maxValue', 1000)}
			progressColor={color('progressColor', '#3EB290')}
			testID="ProgressBar-test-id"
			value={number('value', 1000, { range: true, max: 1000, min: 0, step: 1 })}
		/>
	));
