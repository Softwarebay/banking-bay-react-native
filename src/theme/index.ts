interface Colors {
	screenBackground: string;
	mainDark: string;
	bodyText: string;
	pink: string;
	buttonLight: string;
	buttonDark: string;
	green: string;
	white: string;
	shadow: string;
	grey600: string;
	grey300: string;
	grey100: string;
}

export const colors: Colors = {
	screenBackground: '#FAFCFF',
	mainDark: '#1B1D4D',
	bodyText: '#4C4C60',
	pink: '#FF5887',
	buttonLight: '#97DAFF',
	buttonDark: '#16579D',
	green: '#3EB290',
	white: '#FFFFFF',
	shadow: '#161966',
	grey600: '#868698',
	grey300: '#CED6E1',
	grey100: '#F1F5FD',
};

interface Shadows {
	ios: {
		shadowColor: string;
		shadowOffset: {
			width: number;
			height: number;
		};
		shadowOpacity: number;
		shadowRadius: number;
	};
	android: {
		elevation: number;
		shadowColor: string;
	};
}

export const shadows: Shadows = {
	ios: {
		shadowColor: colors.shadow,
		shadowOffset: {
			width: 0,
			height: 6,
		},
		shadowOpacity: 0.08,
		shadowRadius: 16,
	},
	android: {
		elevation: 4,
		shadowColor: colors.shadow,
	},
};
