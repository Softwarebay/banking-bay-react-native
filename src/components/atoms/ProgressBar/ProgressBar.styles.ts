import { StyleSheet } from 'react-native';
import { colors } from 'theme';

export const styles = StyleSheet.create({
	progress: {
		borderRadius: 10,
		height: 5,
	},
	progressLine: {
		left: 0,
		position: 'absolute',
		top: 0,
	},
	progressShadow: {
		backgroundColor: colors.bodyText,
		opacity: 0.2,
	},
});
