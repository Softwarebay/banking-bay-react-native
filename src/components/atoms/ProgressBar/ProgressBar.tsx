import { StyleProp, View, ViewStyle } from 'react-native';
import { colors } from 'theme';

import { styles } from './ProgressBar.styles';

export interface Props {
	testID: string;
	value: number;
	maxValue: number;
	style?: StyleProp<ViewStyle>;
	progressColor?: string;
}

export const ProgressBar = ({
	maxValue,
	progressColor = colors.green,
	testID,
	value,
	style,
}: Props) => {
	const progress = (value / maxValue) * 100;

	return (
		<View style={style} testID={testID}>
			<View
				style={[
					styles.progress,
					styles.progressLine,
					{
						width: `${progress > 100 ? 100 : progress}%`,
						backgroundColor: progressColor,
					},
				]}
				testID={`progress-line-${testID}`}
			/>
			<View style={[styles.progress, styles.progressShadow]} />
		</View>
	);
};
