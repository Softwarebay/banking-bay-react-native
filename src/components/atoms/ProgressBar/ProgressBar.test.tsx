import { fireEvent, render, screen } from '@testing-library/react-native';
import testRenderer from 'react-test-renderer';

import { colors } from 'theme';
import { ProgressBar } from './ProgressBar';

import type { Props } from './ProgressBar';

describe('ProgressBar', () => {
	const props: Props = {
		maxValue: 100,
		testID: 'progress-bar-test-id',
		value: 80,
	};

	it('should match snapshot', () => {
		const tree = testRenderer.create(<ProgressBar {...props} />).toJSON();

		expect(tree).toMatchSnapshot();
	});

	it('should display green progress line as default color', () => {
		render(<ProgressBar {...props} />);

		const line = screen.getByTestId('progress-line-progress-bar-test-id');

		expect(line).toHaveStyle({
			backgroundColor: colors.green,
		});
	});

	it('should apply other colors to progress line', () => {
		render(<ProgressBar {...props} progressColor="#1ecbe1" />);

		const line = screen.getByTestId('progress-line-progress-bar-test-id');

		expect(line).toHaveStyle({
			backgroundColor: '#1ecbe1',
		});
	});

	it('should calculate progress line width based on value and maxValue', () => {
		render(<ProgressBar {...props} />);

		const line = screen.getByTestId('progress-line-progress-bar-test-id');

		expect(line).toHaveStyle({
			width: '80%',
		});
	});

	it('should display full progress line on progress > 100%', () => {
		render(<ProgressBar {...props} value={300} />);

		const line = screen.getByTestId('progress-line-progress-bar-test-id');

		expect(line).toHaveStyle({
			width: '100%',
		});
	});

	it('should apply additional styles', () => {
		const style = { marginBottom: 10 };
		render(<ProgressBar {...props} style={style} />);

		const bar = screen.getByTestId('progress-bar-test-id');

		fireEvent.press(bar);

		expect(bar).toHaveStyle(style);
	});
});
