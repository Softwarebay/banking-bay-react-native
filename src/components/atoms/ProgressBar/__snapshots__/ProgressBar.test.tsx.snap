// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar should match snapshot 1`] = `
<View
  testID="progress-bar-test-id"
>
  <View
    style={
      Array [
        Object {
          "borderRadius": 10,
          "height": 5,
        },
        Object {
          "left": 0,
          "position": "absolute",
          "top": 0,
        },
        Object {
          "backgroundColor": "#3EB290",
          "width": "80%",
        },
      ]
    }
    testID="progress-line-progress-bar-test-id"
  />
  <View
    style={
      Array [
        Object {
          "borderRadius": 10,
          "height": 5,
        },
        Object {
          "backgroundColor": "#4C4C60",
          "opacity": 0.2,
        },
      ]
    }
  />
</View>
`;
